include/fi_util.h

```diff
@@ -292,6 +293,62 @@ int ofi_get_src_addr(uint32_t addr_format,
 		     const void *dest_addr, size_t dest_addrlen,
 		     void **src_addr, size_t *src_addrlen);
 void ofi_getnodename(char *buf, int buflen);
+int ofi_av_get_index(struct util_av *av, const void *addr);
+
+/*
+ * Connection Map
+ */
+
+enum util_cmap_state {
+	CMAP_UNSPEC,
+	CMAP_CONNECTING,
+	CMAP_CONNECTED
+};
+
+struct util_cmap_handle {
+	struct util_cmap *cmap;
+	enum util_cmap_state state;
+	struct util_cmap_key *key;
+	size_t key_index;
+	fi_addr_t fi_addr;
+	struct util_cmap_peer *peer;
+};
+
+struct util_cmap_key {
+	struct util_cmap_handle *handle;
+};
+DECLARE_FREESTACK(struct util_cmap_key, util_cmap_keypool);
+
+struct util_cmap_peer {
+	struct util_cmap_handle *handle;
+	struct dlist_entry entry;
+	size_t addrlen;
+	uint8_t addr[];
+};
+
+typedef void (*ofi_cmap_free_handle_func)(void *arg);
+
+struct util_cmap {
+	struct util_av *av;
+	struct util_cmap_handle **handles;
+	struct util_cmap_keypool *keypool;
+	struct dlist_entry peer_list;
+	ofi_cmap_free_handle_func free_handle;
+	fastlock_t lock;
+};
+
+void ofi_cmap_update_state(struct util_cmap_handle *handle,
+		enum util_cmap_state state);
+int ofi_cmap_add_handle_addr(struct util_cmap *cmap, struct util_cmap_handle *handle,
```

Comment:
The names of this and some of the following functions are somewhat confusing: it's not easy to tell the real differences between similarly named functions; the "get" function does deletion. If these functions are supposed to be part of the "util" API exposed to the provider writers, please try to refine the names to be more descriptive. Also try to minimize the set of functions need to be exposed. For those functions that require a lock being held, please also add that information to the header file.